libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveDialog.cs

```diff
@@ -464,10 +477,19 @@ protected Task<bool> EndCurrentActionAsync(SequenceContext sequenceContext, Canc
 
         protected async Task<DialogTurnResult> OnEndOfActionsAsync(SequenceContext sequenceContext, CancellationToken cancellationToken = default)
         {
-            // End dialog and return result
+            // Is the current dialog still on the stack?
             if (sequenceContext.ActiveDialog != null)
             {
-                if (ShouldEnd(sequenceContext))
+                // Raise EndOfActions event
+                var endOfActionsEvent = new DialogEvent() { Name = AdaptiveEvents.EndOfActions, Bubble = false };
+                var handled = await OnDialogEventAsync(sequenceContext, endOfActionsEvent, cancellationToken).ConfigureAwait(false);
```

Comment:
This is the lifecycle events discussion.  I just want to think about this a bit more because you could easily end up in an infinite loop.